File: app/src/test/java/com/vrem/wifianalyzer/ActivityUtilsTest.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer;

import android.content.Intent;
import android.view.Window;
import android.view.WindowManager;

import com.vrem.util.IntentUtils;
import com.vrem.wifianalyzer.ActivityUtils.WiFiBandToggle;
import com.vrem.wifianalyzer.navigation.NavigationMenu;
import com.vrem.wifianalyzer.settings.Settings;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.mockStatic;
import static org.powermock.api.mockito.PowerMockito.verifyNoMoreInteractions;
import static org.powermock.api.mockito.PowerMockito.verifyStatic;
import static org.powermock.api.mockito.PowerMockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest(IntentUtils.class)
public class ActivityUtilsTest {

    @Mock
    private Window window;
    @Mock
    private ActionBar actionBar;
    @Mock
    private Toolbar toolbar;
    @Mock
    private Intent intent;

    private MainActivity mainActivity;
    private Settings settings;

    @Before
    public void setUp() {
        mockStatic(IntentUtils.class);

        mainActivity = MainContextHelper.INSTANCE.getMainActivity();
        settings = MainContextHelper.INSTANCE.getSettings();
    }

    @After
    public void tearDown() {
        MainContextHelper.INSTANCE.restore();
        verifyNoMoreInteractions(IntentUtils.class);
        verifyNoMoreInteractions(mainActivity);
        verifyNoMoreInteractions(toolbar);
        verifyNoMoreInteractions(actionBar);
        verifyNoMoreInteractions(window);
        verifyNoMoreInteractions(settings);
        verifyNoMoreInteractions(intent);
    }

    @Test
    public void testSetActionBarOptions() {
        // execute
        ActivityUtils.setActionBarOptions(actionBar);
        // validate
        verify(actionBar).setHomeButtonEnabled(true);
        verify(actionBar).setDisplayHomeAsUpEnabled(true);
    }

    @Test
    public void testSetActionBarOptionsWithNullActionBar() {
        // execute
        ActivityUtils.setActionBarOptions(null);
        // validate
        verify(actionBar, never()).setHomeButtonEnabled(true);
        verify(actionBar, never()).setDisplayHomeAsUpEnabled(true);
    }

    @Test
    public void testSetupToolbar() {
        // setup
        when(mainActivity.findViewById(R.id.toolbar)).thenReturn(toolbar);
        when(mainActivity.getSupportActionBar()).thenReturn(actionBar);
        // execute
        Toolbar actual = ActivityUtils.setupToolbar();
        // validate
        assertEquals(toolbar, actual);

        verify(mainActivity).findViewById(R.id.toolbar);
        verify(mainActivity).getSupportActionBar();

        verify(toolbar).setOnClickListener(any(WiFiBandToggle.class));
        verify(mainActivity).setSupportActionBar(toolbar);
        verify(actionBar).setHomeButtonEnabled(true);
        verify(actionBar).setDisplayHomeAsUpEnabled(true);
    }

    @Test
    public void testWiFiBandToggleOnClickToggles() {
        // setup
        when(mainActivity.getCurrentNavigationMenu()).thenReturn(NavigationMenu.CHANNEL_GRAPH);
        WiFiBandToggle wiFiBandToggle = new WiFiBandToggle(mainActivity);
        // execute
        wiFiBandToggle.onClick(null);
        // validate
        verify(settings).toggleWiFiBand();
        verify(mainActivity).getCurrentNavigationMenu();
    }

    @Test
    public void testWiFiBandToggleOnClickDoesNotToggles() {
        // setup
        when(mainActivity.getCurrentNavigationMenu()).thenReturn(NavigationMenu.ACCESS_POINTS);
        WiFiBandToggle wiFiBandToggle = new WiFiBandToggle(mainActivity);
        // execute
        wiFiBandToggle.onClick(null);
        // validate
        verify(settings, never()).toggleWiFiBand();
        verify(mainActivity).getCurrentNavigationMenu();
    }

    @Test
    public void testKeepScreenOnSwitchOn() {
        // setup
        when(settings.isKeepScreenOn()).thenReturn(true);
        when(mainActivity.getWindow()).thenReturn(window);
        // execute
        ActivityUtils.keepScreenOn();
        // validate
        verify(settings).isKeepScreenOn();
        verify(mainActivity).getWindow();
        verify(window).addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    @Test
    public void testKeepScreenOnSwitchOff() {
        // setup
        when(settings.isKeepScreenOn()).thenReturn(false);
        when(mainActivity.getWindow()).thenReturn(window);
        // execute
        ActivityUtils.keepScreenOn();
        // validate
        verify(settings).isKeepScreenOn();
        verify(mainActivity).getWindow();
        verify(window).clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    @Test
    public void testStartWiFiSettings() {
        // setup
        when(IntentUtils.makeIntent(android.provider.Settings.Panel.ACTION_WIFI)).thenReturn(intent);
        // execute
        ActivityUtils.startWiFiSettings();
        // validate
        verify(mainActivity).startActivityForResult(intent, 0);
        verifyStatic(IntentUtils.class);
        IntentUtils.makeIntent(android.provider.Settings.Panel.ACTION_WIFI);
    }

    @Test
    public void testStartLocationSettings() {
        // setup
        when(IntentUtils.makeIntent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS)).thenReturn(intent);
        // execute
        ActivityUtils.startLocationSettings();
        // validate
        verify(mainActivity).startActivity(intent);
        verifyStatic(IntentUtils.class);
        IntentUtils.makeIntent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
    }

}

File: app/src/test/kotlin/com/vrem/wifianalyzer/ActivityUtilsTest.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2026 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer

import android.content.Intent
import android.os.Build
import android.provider.Settings
import android.view.View
import android.view.Window
import android.view.WindowManager
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import org.robolectric.annotation.Config

@RunWith(AndroidJUnit4::class)
@Config(sdk = [Build.VERSION_CODES.BAKLAVA])
class ActivityUtilsTest {
    private val window: Window = mock()
    private val actionBar: ActionBar = mock()
    private val toolbar: Toolbar = mock()
    private val intent: Intent = mock()
    private val intentArgumentCaptor = argumentCaptor<Intent>()
    private val mainActivity = MainContextHelper.INSTANCE.mainActivity
    private val settings = MainContextHelper.INSTANCE.settings

    @After
    fun tearDown() {
        MainContextHelper.INSTANCE.restore()
        verifyNoMoreInteractions(mainActivity)
        verifyNoMoreInteractions(toolbar)
        verifyNoMoreInteractions(actionBar)
        verifyNoMoreInteractions(window)
        verifyNoMoreInteractions(settings)
        verifyNoMoreInteractions(intent)
    }

    @Test
    fun setupToolbar() {
        // setup
        doReturn(toolbar).whenever(mainActivity).findViewById<View>(R.id.toolbar)
        doReturn(actionBar).whenever(mainActivity).supportActionBar
        // execute
        val actual = mainActivity.setupToolbar()
        // validate
        assertThat(actual).isEqualTo(toolbar)
        verify(mainActivity).findViewById<View>(R.id.toolbar)
        verify(mainActivity).supportActionBar
        verify(mainActivity).setSupportActionBar(toolbar)
        verify(actionBar).setHomeButtonEnabled(true)
        verify(actionBar).setDisplayHomeAsUpEnabled(true)
    }

    @Test
    fun keepScreenOnSwitchOn() {
        // setup
        doReturn(true).whenever(settings).keepScreenOn()
        doReturn(window).whenever(mainActivity).window
        // execute
        mainActivity.keepScreenOn()
        // validate
        verify(settings).keepScreenOn()
        verify(mainActivity).window
        verify(window).addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    @Test
    fun keepScreenOnSwitchOff() {
        // setup
        doReturn(false).whenever(settings).keepScreenOn()
        doReturn(window).whenever(mainActivity).window
        // execute
        mainActivity.keepScreenOn()
        // validate
        verify(settings).keepScreenOn()
        verify(mainActivity).window
        verify(window).clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    @Test
    fun startWiFiSettings() {
        // execute
        mainActivity.startWiFiSettings()
        // validate
        verify(mainActivity).startActivity(intentArgumentCaptor.capture())
        assertThat(intentArgumentCaptor.firstValue.action).isEqualTo(Settings.Panel.ACTION_WIFI)
    }

    @Test
    fun startLocationSettings() {
        // execute
        mainActivity.startLocationSettings()
        // validate
        verify(mainActivity).startActivity(intentArgumentCaptor.capture())
        assertThat(intentArgumentCaptor.firstValue.action).isEqualTo(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
    }
}
